node/node.go

```diff
@@ -406,42 +420,95 @@ func updateNodeLease(ctx context.Context, leases v1beta1.LeaseInterface, lease *
-func patchNodeStatus(nodes v1.NodeInterface, nodeName types.NodeName, oldNode *corev1.Node, newNode *corev1.Node) (*corev1.Node, []byte, error) {
-	patchBytes, err := preparePatchBytesforNodeStatus(nodeName, oldNode, newNode)
-	if err != nil {
-		return nil, nil, err
+func prepareThreewayPatchBytesForNodeStatus(nodeFromProvider, apiServerNode *corev1.Node) ([]byte, error) {
+	// We use these two values to calculate a patch. We use a three-way patch in order to avoid
+	// causing state regression server side. For example, let's consider the scenario:
+	/*
+		+-----+                                           +-----+                              +-----+
+		| VK  |                                           | K8s |                              | Ext |
+		+-----+                                           +-----+                              +-----+
+		   | ---------------------------------------------\  |                                    |
+		   |-| Updates internal node conditions to [A, B] |  |                                    |
+		   | |--------------------------------------------|  |                                    |
+		   |                                                 |                                    |
+		   | Patch upsert: [A, B]                            |                                    |
+		   |------------------------------------------------>|                                    |
+		   |                  -----------------------------\ |                                    |
+		   |                  | Node conditions are [A, B] |-|                                    |
+		   |                  |----------------------------| |                                    |
+		   |                                                 |                                    |
+		   |                                                 |                    Patch upsert: C |
+		   |                                                 |<-----------------------------------|
+		   |                                                 | --------------------------------\  |
+		   |                                                 |-| Node conditions are [A, B, C] |  |
+		   |                                                 | |-------------------------------|  |
+		   | ------------------------------------------\     |                                    |
+		   |-| Updates internal node conditions to [A] |     |                                    |
+		   | |-----------------------------------------|     |                                    |
+		   |                                                 |                                    |
+		   | Patch delete: B, upsert A                       |                                    |
+		   |------------------------------------------------>|                                    |
+		   |                                                 |                                    |
```

Comment:
I think a few lines are missing here clearly showing the problem is that node conditions are `[A]` instead of `[A, C]`, right?